engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala

```diff
@@ -0,0 +1,148 @@
+package cromwell.engine.workflow.lifecycle.execution.callcaching
+
+import java.nio.file.Path
+
+import akka.actor.{ActorLogging, ActorRef, LoggingFSM}
+import cromwell.backend.BackendJobDescriptor
+import cromwell.engine.workflow.lifecycle.execution.callcaching.EngineJobHashingActor._
+import cromwell.engine.workflow.lifecycle.execution.callcaching.FileHasherActor.JobFileHashRequests
+
+/**
+  * The EJHA, pronounced "edge-ha".
+  *
+  * Over time will emit up to two messages back to its parent:
+  *  * (if read enabled): Either a CacheHit(id) or CacheMiss message
+  *  * (if write enabled): A CallCacheHashes(hashes) message
+  */
+case class EngineJobHashingActor(jobDescriptor: BackendJobDescriptor,
+                                 fileHasherActor: ActorRef,
+                                 mode: CallCachingMode) extends LoggingFSM[EJHAState, EJHAData] with ActorLogging {
+
+  val receiver = context.parent
+
+  initialise()
+
+  when(DeterminingHitOrMiss) {
+    case Event(hashResultMessage: HashResultMessage, _) => findRelevantCacheResultsThenTransition(hashResultMessage.hashes)
+    case Event(newCacheResults: CacheResultMatchesForHashes, _) => checkWhetherHitOrMissIsKnownThenTransition(stateData.intersectCacheResultIds(newCacheResults))
+  }
+
+  when(GeneratingAllHashes) {
+    case Event(hashResultMessage: HashResultMessage, _) => checkWhetherAllHashesAreKnownAndTransition(stateData.withNewKnownHashes(hashResultMessage.hashes))
+    case Event(CacheResultMatchesForHashes(hashResults, cacheMatches), _) => stay // Don't care; we already know the hit/miss status. Ignore this message
```

Comment:
If I've read the code correctly, when this runs in `WriteCache` we won't actually know (or care) about hit/miss status, so the comment seems a bit misleading.